modules/currency-l0/src/main/scala/org/tessellation/currency/l0/snapshot/services/StateChannelBinarySender.scala

```diff
@@ -0,0 +1,200 @@
+
+      private def updateExpectedOrdinalsToConfirm(
+        confirmedTracked: Queue[TrackedBinary],
+        snapshotOrdinal: SnapshotOrdinal
+      ): F[PosLong] =
+        expectedOrdinalsToConfirmR.updateAndGet { prev =>
+          if (confirmedTracked.isEmpty) prev
+          else {
+            val ordinalsPassed = confirmedTracked.map(tb => snapshotOrdinal.value - tb.sentAtOrdinal.value)
+            val avgOrdinalsPassed = PosLong.from(Math.round(ordinalsPassed.sum.toDouble / ordinalsPassed.length)).getOrElse(prev)
+            avgOrdinalsPassed
+          }
+        }
+
+      private def updateAvgConfirmedCountPerSnapshot(confirmedTracked: Queue[TrackedBinary]): F[Unit] =
+        avgConfirmedCountPerSnapshotR.update { prev =>
+          NonNegLong.from {
+            Math.ceil((prev.toDouble + confirmedTracked.length.toDouble) / 2.0).toLong
+          }.getOrElse(prev)
+        }
+
+      private def updateCatchUpMode(
+        pending: Queue[TrackedBinary],
+        snapshotOrdinal: SnapshotOrdinal,
+        expectedOrdinalsToConfirm: PosLong
+      ): F[Unit] =
+        catchUpMode.update(prev =>
+          if (pending.isEmpty) {
+            false
+          } else {
+            val hasStalled = pending.exists { p =>
+              val ordinalsWithoutConfirmationCount =
+                NonNegLong
+                  .from(snapshotOrdinal.value - p.sentAtOrdinal.value)
+                  .getOrElse(NonNegLong.MaxValue)
+              ordinalsWithoutConfirmationCount >= expectedOrdinalsToConfirm
+            }
+            hasStalled || prev
+          }
+        )
+
+      private def enqueue(binary: Hashed[StateChannelSnapshotBinary]): F[Unit] =
+        for {
+          currentOrdinal <- lastGlobalSnapshotStorage.getOrdinal.map(_.getOrElse(SnapshotOrdinal.MinValue))
+          _ <- pendingR.update { pending =>
+            pending.enqueue(TrackedBinary(binary, currentOrdinal))
+          }
+        } yield ()
+
+      private def post(binary: Hashed[StateChannelSnapshotBinary]): F[Unit] = {
+        val sendRetries = 5
+
+        val retryPolicy: RetryPolicy[F] = RetryPolicies.limitRetries(sendRetries)
+
+        def wasSuccessful: Either[NonEmptyList[StateChannelValidationError], Unit] => F[Boolean] =
+          _.isRight.pure[F]
+
+        def onFailure(binaryHashed: Hashed[StateChannelSnapshotBinary]) =
+          (_: Either[NonEmptyList[StateChannelValidationError], Unit], details: RetryDetails) =>
+            logger.info(s"Retrying sending ${binaryHashed.hash.show} to Global L0 after rejection. Retries so far ${details.retriesSoFar}")
+
+        def onError(binaryHashed: Hashed[StateChannelSnapshotBinary]) = (_: Throwable, details: RetryDetails) =>
+          logger.info(s"Retrying sending ${binaryHashed.hash.show} to Global L0 after error. Retries so far ${details.retriesSoFar}")
+
+        retryingOnFailuresAndAllErrors[Either[NonEmptyList[StateChannelValidationError], Unit]](
```

Comment:
`retryingOnAllErrors` works with `MonadError` so compiler forces me to provide it implicitly for request wrapped by `F`
`retryingOnFailures` works with `wasSuccessful` predicate so we can decide if retry or not

`retryingOnFailuresAndAllErrors` is a combination of both so as long as `java.lang.Error` thrown in the request is wrapped by `MonadError` then it will be catched. All the operations inside retry policy are wrapped by `F` with `MonadError[F, E]` so we should be fine.

Reply:
That's my concern, that we're catching `java.lang.Error` where my understanding is almost all cases should not be caught.

Reply:
Domain-specific errors are exposed as `Either` and `StateChannelValidationError`  and handled by `retryingOnFailures`. The other errors are catched by `retryingOnAllErrors`.

If we only want to catch our errors then we could switch from `retryingOnFailuresAndAllErrors` to `retryingOnFailures` but for me, the current implementation is fine.

Reply:
Also, keep in mind that we also use `retryingOnAllErrors` policy in consensus and healthcheck so if we want to switch then we should switch everywhere.

Reply:
@marcinwadon @TheMMaciek wdyt? We use various retry policies in our code. Some are retrying only on domain-specific `Throwable`, some on `Left`, some on any `Throwable`s.

Reply:
@kpudlik Ok, I think I've answered my own question and now I remember looking into this before.

I think we're ok because Cats has checks for `NonFatal` exceptions so it looks like those are the only ones that will be trapped.